zkevm-circuits/src/pi_circuit2.rs

```diff
@@ -1044,199 +948,215 @@ impl<F: Field> SubCircuitConfig<F> for PiCircuitConfig<F> {
+                |cb| {
+                    let length = meta.query_advice(blk_hdr_rlp_len_calc, Rotation(32));
+                    cb.require_equal("RLP var field", byte.expr(), 0x80.expr() + length.expr());
+                },
             );
 
+            // Check total length of RLP stream
             // For the block header, the total RLP length is always two bytes long and only
             // the LSB fluctuates: Minimum total length: lengths of all the
             // fixed size fields + all the RLP headers = 527 bytes (0x020F)
             // Maximum total length: minimum total length + (maximum length of variable zize
             // field) = 527 + 4*32+1*8 = 663 (0x0297) Actual total length:
             // minimum total length + length of all variable size fields (number, gas_limit,
             // gas_used, timestamp, base fee)
-            cb.require_equal(
-                "LSB(total_len) = min(LSB(total_len)) + sum(Len(<var field>))",
-                total_len,
-                0x0F.expr()
-                    + number_len
-                    + gas_limit_len
-                    + gas_used_len
-                    + timestamp_len
-                    + base_fee_len,
-            );
-
-            cb.gate(meta.query_selector(q_blk_hdr_total_len))
-        });
-
-        // Reconstruct field values
-        meta.create_gate("Block header RLP: Calculate fields' value RLC", |meta| {
-            let mut cb = BaseConstraintBuilder::new(MAX_DEGREE);
-
-            let q_reconstruct_cur = meta.query_fixed(q_reconstruct, Rotation::cur());
-            let q_reconstruct_next = meta.query_fixed(q_reconstruct, Rotation::next());
-            let blk_hdr_rlp_next = meta.query_advice(blk_hdr_rlp, Rotation::next());
-            let blk_hdr_reconstruct_value_cur =
-                meta.query_advice(blk_hdr_reconstruct_value, Rotation::cur());
-            let blk_hdr_reconstruct_value_next =
-                meta.query_advice(blk_hdr_reconstruct_value, Rotation::next());
+            cb.condition(q_total_length, |cb| {
+                let mut get_len = |offset: usize| {
+                    meta.query_advice(
+                        blk_hdr_rlp_len_calc,
+                        Rotation((offset - 3).try_into().unwrap()),
+                    )
+                };
+                let number_len = get_len(NUMBER_RLP_OFFSET + NUMBER_SIZE);
+                let gas_limit_len = get_len(GAS_LIMIT_RLP_OFFSET + GAS_LIMIT_SIZE);
+                let gas_used_len = get_len(GAS_USED_RLP_OFFSET + GAS_USED_SIZE);
+                let timestamp_len = get_len(TIMESTAMP_RLP_OFFSET + TIMESTAMP_SIZE);
+                let base_fee_len = get_len(BASE_FEE_RLP_OFFSET + BASE_FEE_SIZE);
+                cb.require_equal(
+                    "total_len",
+                    byte.expr(),
+                    0x0F.expr() // TODO(Brecht): With the explanation above I would expect this to be 0x020F
```

Comment:
The total length can be  from 0x020F to 0x0297. The 0x02 part never changes thus it is checked as a constant.